components/brave_wallet/browser/eth_logs_tracker.h

```diff
@@ -0,0 +1,65 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_ETH_LOGS_TRACKER_H_
+#define BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_ETH_LOGS_TRACKER_H_
+
+#include <string>
+#include <vector>
+
+#include "base/memory/raw_ptr.h"
+#include "base/observer_list.h"
+#include "base/observer_list_types.h"
+#include "base/time/time.h"
+#include "base/timer/timer.h"
+#include "brave/components/brave_wallet/browser/json_rpc_service.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "brave/components/brave_wallet/common/brave_wallet_types.h"
+
+namespace brave_wallet {
+
+class JsonRpcService;
+
+class EthLogsTracker {
+ public:
+  explicit EthLogsTracker(JsonRpcService* json_rpc_service);
+  EthLogsTracker(const EthLogsTracker&) = delete;
+  EthLogsTracker& operator=(const EthLogsTracker&) = delete;
+  EthLogsTracker(const EthLogsTracker&&) = delete;
+  EthLogsTracker& operator=(const EthLogsTracker&&) = delete;
+
+  virtual ~EthLogsTracker();
+
+  class Observer : public base::CheckedObserver {
+   public:
+    virtual void OnLogsReceived(base::Value rawlogs) = 0;
+  };
+
+  // If timer is already running, it will be replaced with new interval
+  void Start(base::TimeDelta interval);
+  void Stop();
+  bool IsRunning() const;
+
+  void AddObserver(Observer* observer);
+  void RemoveObserver(Observer* observer);
+
+ private:
+  void GetLogs();
+  void OnGetLogs(const std::vector<Log>& logs,
+                 base::Value rawlogs,
+                 mojom::ProviderError error,
+                 const std::string& error_message);
+
+  base::RepeatingTimer timer_;
+  raw_ptr<JsonRpcService> json_rpc_service_ = nullptr;
+
+  base::ObserverList<Observer> observers_;
+
+  base::WeakPtrFactory<EthLogsTracker> weak_factory_;
```

Comment:
Common pattern to initialize as `weak_factory_{this};` instead of in constructor.

Reply:
fixed